src/main/java/htsjdk/samtools/SAMFileWriterImpl.java

```diff
@@ -243,9 +245,22 @@ public final void close()
     /**
      * Write the header to disk.  Header object is available via getHeader().
      * @param textHeader for convenience if the implementation needs it.
+     * @deprecated since 06/2017. {@link #writeHeader(Supplier)} is preferred for avoid String construction if not need it.
      */
+    @Deprecated
     abstract protected void writeHeader(String textHeader);
 
+    /**
+     * Write the header to disk. Header object is available via getHeader().
+     *
+     * <p>Note: default implementation uses {@link #writeHeader(String)}.
+     *
+     * @param textHeaderSupplier for convenience if the implementation needs it.
+     */
+    protected void writeHeader(Supplier<String> textHeaderSupplier) {
```

Comment:
This seems like a bit of a confusing mechanism for allowing people to avoid converting the header to string.  Would it make more sense to add a `writeHeader(SAMFileHeader header) method that takes SAMFileHeader and constructs the String from it?  Then provide an implementation of it that does the conversion to string.  Your class can  use that override and not write anything. The `CRAMFileWriter` could make use of that method as well, since it currently takes the header string, converts it back to a header and then must convert that again.

Reply:
Sounds good. Done and modify the rest of implementations to move away from the `String` method.